sdks/python/apache_beam/examples/cookbook/mergecontacts_test.py

```diff
@@ -101,22 +99,61 @@ def normalize_tsv_results(self, tsv_data):
+    p = beam.Pipeline('DirectPipelineRunner')
+
+    contacts_email = p | 'create_email' >>  beam.Create([self.CONTACTS_EMAIL])
+    contacts_phone = p | 'create_phone' >> beam.Create([self.CONTACTS_PHONE])
+    contacts_snailmail = (p | 'create_snail_mail' >>
+                          beam.Create([self.CONTACTS_SNAILMAIL]))
+
+    email = (contacts_email
+             | beam.Map('backslash_email', lambda x: re.sub(r'\\', r'\\\\', x))
+             | beam.Map('escape_quotes_email', lambda x: re.sub(r'"', r'\"', x))
+             | beam.Map('split_email', lambda x: re.split(r'\t+', x, 1)))
+    phone = (contacts_phone
+             | beam.Map('backslash_phone', lambda x: re.sub(r'\\', r'\\\\', x))
+             | beam.Map('escape_quotes_phone', lambda x: re.sub(r'"', r'\"', x))
+             | beam.Map('split_phone', lambda x: re.split(r'\t+', x, 1)))
+    snailmail = (contacts_snailmail
+                 | beam.Map('backslash_snailmail',
+                            lambda x: re.sub(r'\\', r'\\\\', x))
+                 | beam.Map('escape_quotes_snailmail',
+                            lambda x: re.sub(r'"', r'\"', x))
+                 | beam.Map('split_snailmail',
+                            lambda x: re.split(r'\t+', x, 1)))
+
+    grouped = (email, phone, snailmail) | 'group_by_name' >> beam.CoGroupByKey()
+
+    result_tsv_lines = (grouped | 'result_tsv' >> beam.Map(
+        lambda (name, (email, phone, snailmail)): '\t'
+        .join(['"%s"' % name,
+               '"%s"' % ','.join(sorted(email.strip('"').split(','))),
+               '"%s"' % ','.join(sorted(phone.strip('"').split(','))),
+               '"%s"' % next(iter(snailmail), '')])))
+
+    luddites = (grouped | 'filter_luddites' >> beam.Filter(
+        lambda (name, (email, phone, snailmail)
+               ): not next(iter(email), None)))
+    writers = (grouped | 'filter_writers' >> beam.Filter(
+        lambda (name, (email, phone, snailmail)
+               ): not next(iter(phone), None)))
+    nomads = (grouped | 'filter_nomads' >> beam.Filter(
+        lambda (name, (email, phone, snailmail)
+               ): not next(iter(snailmail), None)))
+
+    num_luddites = luddites | 'luddites' >> beam.combiners.Count.Globally()
+    num_writers = writers | 'writers' >> beam.combiners.Count.Globally()
+    num_nomads = nomads | 'nomads' >> beam.combiners.Count.Globally()
+    #(self.normalize_tsv_results('\n'.join(result_tsv_lines)),
+    beam.assert_that(result_tsv_lines,
+                     beam.equal_to([self.EXPECTED_TSV]),
+                     label='assert:tag_tsv_results')
+    beam.assert_that(num_luddites, beam.equal_to([self.EXPECTED_LUDDITES]),
+                     label='assert:tag_num_luddites')
+    beam.assert_that(num_writers, beam.equal_to([self.EXPECTED_WRITERS]),
+                     label='assert:tag_num_writers')
+    beam.assert_that(num_nomads, beam.equal_to([self.EXPECTED_NOMADS]),
+                     label='assert:tag_num_nomads')
 
```

Comment:
Always call p.run() to run the asserts. 

Alternatively, do
```
with Pipeline(...) as p:
    ...
```

which calls run on exiting the block.